src/machines/java/com/enderio/machines/common/io/fluid/MachineFluidTank.java

```diff
@@ -0,0 +1,250 @@
+package com.enderio.machines.common.io.fluid;
+
+import com.enderio.EnderIO;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.BucketItem;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraftforge.common.capabilities.ForgeCapabilities;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidType;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidHandlerItem;
+import net.minecraftforge.fluids.capability.templates.FluidTank;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
```

Comment:
So what exactly is this class for? It does something like a fluid bar, and some things for in world interactions and fluidutil? I seemed to have glossed over this when I first looked, but is this all needed like this?

Reply:
The MachineFluidTank is a class that does similar things to the FluidTank class that Forge provides, it holds data about what type of fluid is contained and how much. It has a capacity and an API for transferring into and out of the tank. The Forge implementation is missing filters for controlling which fluids are allowed to enter and exit the tank. My implementation allows for flagging a tank as for example an input tank accepting only fluids that has the coolant property set for example. It would reject any attempt from external sources to pull coolant out or put lava into the tank. As for the enchantment infuser or whatever it's called you flag it as an input tank and set the predicate to only accept fluidXP and only do it when the machine has a recipe, also set the capacity to whatever amount the recipe requires. This will disallow any fluid from going into the tank until the machine has a valid recipe. Then it only accepts liqudXP and only the amount specified by the recipe. It won't allow liquidXP to be pulled out by pipes, however when you are done with your recipe and want to void the fluid that went into the recipe you just use tank.drain([tank.get_amount()], true) to force it to drain all internal content. The reason to why it is so long is that I had to rewrite like the entire class to enable the predicate and to remove the functionality of storing multiple fluids (like in the tinkers smeltery). If you want to I'll point you towards my unfinished implementation of the vat, but it is still missing all the recipe stuff.